protocol/consistencychecks.go

```diff
@@ -225,43 +229,65 @@ func (cc *ConsistencyChecks) verifySTR(str *m.SignedTreeRoot) error {
 	// verify hash chain
 	// TODO: clearly verify which is the client's actual expectation.
 	// See #81
-	if (str.Epoch == cc.CurrentEpoch && bytes.Equal(cc.SavedSTR, str.Signature)) ||
-		(str.Epoch == cc.CurrentEpoch+1 && str.VerifyHashChain(cc.SavedSTR)) {
+	if (str.Epoch == cc.SavedSTR.Epoch && bytes.Equal(cc.SavedSTR.Signature, str.Signature)) ||
+		(str.Epoch == cc.SavedSTR.Epoch+1 && str.VerifyHashChain(cc.SavedSTR.Signature)) {
 		return nil
 	}
+
+	// TODO: verify the directory's policies as well. See #115
 	return ErrorBadSTR
 }
 
-// verifySignedPromise checks if the returned response is match
-// with the directory's policy. Then it calls verifyFulfilledPromise
-// to verify whether issued TBs was inserted in the directory
-// as promised. Finally, it verifies the returned promise
-// based on the request type. Note that the directory returns a promise
-// iff the returned proof is a proof of absence.
+// verifyFulfilledPromise verifies issued TBs was inserted
+// in the directory as promised.
+func (cc *ConsistencyChecks) verifyFulfilledPromise(uname string,
+	str *m.SignedTreeRoot,
+	ap *m.AuthenticationPath) error {
+	if !cc.useTBs {
+		return nil
+	}
+
+	for name, tb := range cc.TBs {
+		if cc.SavedSTR.Epoch+1 == str.Epoch &&
+			name == uname {
+			if !tb.Verify(ap) {
+				return ErrorBrokenPromise
+			}
+			delete(cc.TBs, uname)
+			return nil
+		} else if cc.SavedSTR.Epoch < str.Epoch {
+			// TODO: should verify when we introduce KeyLookupInEpoch
+			// and monitoring with missed epochs,
+			// see: https://github.com/coniks-sys/coniks-go/pull/74#discussion_r84930999
+			// and: https://github.com/coniks-sys/coniks-go/pull/74#discussion_r84501211
+			delete(cc.TBs, name)
+		}
+		// keep current epoch's returned promises
+		// for future verifications
+	}
+	return nil
```

Comment:
> I realized that how the client could know whether it expects to have a cached TB or not? Maybe something like a map `cachedTBs: map[username]epochToBeVerified` (or something else, but it must be controlled by the client)?

Right, good point. The client definitely needs a way to keep track of when it receives each TB, but I think it's ok for the client to control it since it already needs to keep track of TBs.

> If the client expects to have a cached TB and it doesn't have one when `verifyFulfilledPromise`, maybe the reason could be the client somehow lost that cached TB in `cc.TBs` (e.g., by mistaken implementation etc), then the same issue could happen to `cachedTBs` map, too.

This is totally true, so I'm not sure that having a second map `cachedTBs` is the best solution. Is it possible for `cc.TBs` to store `(tb, epoch)` tuples? Or maybe better, store `(tb, str)` tuples?